# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: js/src/vm/Shape.cpp
# Commit: 3be60f42358a
# Full Hash: 3be60f42358afa252f4e8fba01baba156dbd7ffe
# Author: Jan de Mooij <jdemooij@mozilla.com>
# Date: 2021-03-24 08:48:48
# Description:
#   Bug 1700525 - Set object flags on the last property's shape instead of on the property's shape. r=iain
#   
#   This fixes a regression from bug 1696178 part 5 with dictionary objects.
#   
#   Object flags on non-last-properties aren't used anywhere, we need to set them on the
# ==============================================================================

diff -r 75b000d18d03 -r 3be60f42358a js/src/vm/Shape.cpp
--- a/js/src/vm/Shape.cpp	Wed Mar 24 05:15:11 2021 +0000
+++ b/js/src/vm/Shape.cpp	Wed Mar 24 05:51:01 2021 +0000
@@ -1009,8 +1009,12 @@
       }
     }
 
+    // Update the last property's object flags. This is fine because we just
+    // generated a new shape for the object. Note that |shape|'s object flags
+    // aren't used anywhere if it's not the last property.
+    obj->lastProperty()->setObjectFlags(objectFlags);
+
     shape->setBase(obj->lastProperty()->base());
-    shape->setObjectFlags(objectFlags);
     shape->setSlot(slot);
     shape->attrs = uint8_t(attrs);
     shape->immutableFlags &= ~Shape::ACCESSOR_SHAPE;
@@ -1031,6 +1035,8 @@
     }
   }
 
+  MOZ_ASSERT(obj->lastProperty()->objectFlags() == objectFlags);
+
   MOZ_ASSERT(shape->isDataProperty());
   return shape;
 }
@@ -1106,8 +1112,12 @@
       return nullptr;
     }
 
+    // Update the last property's object flags. This is fine because we just
+    // generated a new shape for the object. Note that |shape|'s object flags
+    // aren't used anywhere if it's not the last property.
+    obj->lastProperty()->setObjectFlags(objectFlags);
+
     shape->setBase(obj->lastProperty()->base());
-    shape->setObjectFlags(objectFlags);
     shape->setSlot(SHAPE_INVALID_SLOT);
     shape->attrs = uint8_t(attrs);
     shape->immutableFlags |= Shape::IN_DICTIONARY | Shape::ACCESSOR_SHAPE;
@@ -1143,6 +1153,8 @@
     obj->freeSlot(cx, oldSlot);
   }
 
+  MOZ_ASSERT(obj->lastProperty()->objectFlags() == objectFlags);
+
   MOZ_ASSERT(!shape->isDataProperty());
   return shape;
 }
